### Installing and Importing SQL libraries

In [1]:
#pip install psycopg2
#pip install ipython-sql
#pip install --upgrade ipython

import psycopg2
%load_ext sql

### Connecting to Database Postgre

In [2]:
%sql postgresql://ugevpyllcosyvj:c6d71dd4ed2886ccd8eb2286879a05dc3ebb21046ff280be09bff424ec3d15ff@ec2-44-198-100-81.compute-1.amazonaws.com:5432/db4vmf7pt6qmqj
# +psycopg2

### Get to know the Data

In [3]:
%%sql 
select *
from information_schema.tables
where table_schema = 'public'

 * postgresql://ugevpyllcosyvj:***@ec2-44-198-100-81.compute-1.amazonaws.com:5432/db4vmf7pt6qmqj
5 rows affected.


[('db4vmf7pt6qmqj', 'public', 'dim_geography', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('db4vmf7pt6qmqj', 'public', 'dim_product', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('db4vmf7pt6qmqj', 'public', 'fact_sales', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('db4vmf7pt6qmqj', 'public', 'dim_territory', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('db4vmf7pt6qmqj', 'public', 'dim_customer', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None)]

### Question 1

Untuk mengetahui target customer BeeCycle. Kamu melakukan query untuk melihat distribusi customer berdasarkan gender dan umur.

Dimana untuk category umur, kamu membagi umur customer kedalam (Hint : CASE WHEN)

    umur customer <= 20 tahun maka 'Group <=20'
    umur customer antara 21 dan 40 tahun maka 'Group 21 - 40'
    umur customer antara 41 dan 60 tahun maka 'Group 41 - 60'
    umur customer lebih dari 60 tahun maka 'Group >60'

Dengan grouping_age yg sudah didefine, selanjutnya jawab pertanyaan berikut

Grouping age apa dan gender apa yg memiliki transaksi paling tinggi di BeeCycle?

In [4]:
%%sql
select gender,
case when (date_part('year', current_date) - date_part('year', birthdate)) <= 20 then 'Group <=20' 
when (date_part('year', current_date) - date_part('year', birthdate)) between 21 and 40 then 'Group 21 - 40'
when (date_part('year', current_date) - date_part('year', birthdate)) between 41 and 60 then 'Group 41 - 60'
else 'Group > 60' end Age_category,
count(order_detail_id) sales
from dim_customer
inner join fact_sales using(customer_id)
group by 1,2
order by sales desc

 * postgresql://ugevpyllcosyvj:***@ec2-44-198-100-81.compute-1.amazonaws.com:5432/db4vmf7pt6qmqj
8 rows affected.


[('F', 'Group 21 - 40', 1613),
 ('M', 'Group 21 - 40', 1454),
 ('F', 'Group 41 - 60', 1395),
 ('M', 'Group 41 - 60', 1254),
 ('M', 'Group > 60', 148),
 ('F', 'Group > 60', 72),
 ('M', 'Group <=20', 10),
 ('F', 'Group <=20', 7)]

Grouping Age yang memiliki transaksi terbanyak di BeeCycle adalah rentang 21 sampai 40 tahun diikuti oleh gouping age 41 sampai 60 tahun di posisi ke 2, terakhir diikuti oleh grouping age di atas 60 tahun.
Dengan rincian, perempuan grouping age 21 sampai 40 tahun memiliki jumlah transaksi sebesar 1613 kali transaksi, diikuti oleh laki-laki di group age yang sama sebanyak 1454 kali transaksi.

### Question 2

Karena bulan Oktober depan akan ada campaign. Kamu punya ide untuk menghilight warna product tertentu yg menjadi kegemaran customer. Kamu selanjutnya melakukan query untuk menjawab pertanyaan berikut :

Warna apa di tiap tahun yg menjadi warna paling populer dibeli oleh customer?

In [5]:
%%sql
with CTE1 as
(select color,
extract('year' from order_date) years, 
count(order_detail_id) sales,
rank() over (partition by extract('year' from order_date) order by count(distinct order_detail_id) desc) sort
from fact_sales
left join dim_product using(product_id)
group by 1,2
order by 2,3)

select color,years,sales
from CTE1
where sort = 1

 * postgresql://ugevpyllcosyvj:***@ec2-44-198-100-81.compute-1.amazonaws.com:5432/db4vmf7pt6qmqj
6 rows affected.


[('Red', 2016.0, 91),
 ('Red', 2017.0, 183),
 ('NA', 2018.0, 649),
 ('NA', 2019.0, 1010),
 ('NA', 2020.0, 540),
 ('NA', 2021.0, 156)]

Warna produk paling laris di BeeCycle per tahun dalam 6 tahun terakhir: <br>
Pada tahun 2016 produk warna merah terjual sebanyak 91 produk.<br>
Tahun 2017 masih dengan warna yang sama, terjual sebanyak 183 produk. <br>
Mulai tahun 2018 NA menggeser merah sebagai warna paling laris terjual sebanyak 649 produk. <br>
Tahun 2019 NA masih menjadi warna paling laris diantara pembeli, terjual sebanyak 1010 produk. <br>
Tahun 2020 masih dengan warna yang sama terjual sebanyak 540 produk. <br>
Dan tahun 2021 NA masih warna paling laris diantara pembeli, namun tahun ini tercatat sebagai penjualan warna NA paling sedikit dari tahun 2018 terjual sebanyak 156 produk.

### Question 3

Karena campaign bulan Oktober tsb hanya terbatas untuk tiap2 toko cabang. kamu punya ide, bagaimana hanya mengambil TOP 10 product_id dari masing2 toko cabang. sehingga kamu melakukan query untuk menjawab pertanyaan berikut :

TOP 10 product name apa saja yg paling populer dari masing-masing territory>

In [6]:
%%sql
with CTE2 as
(select product_name,
territory_id,
count(order_detail_id) sales,
rank() over (partition by territory_id  order by count(distinct order_detail_id) desc) sort
from fact_sales
left join dim_product using(product_id)
group by 1,2
order by 2,3 desc)

select product_name,territory_id,sales
from CTE2
where sort <= 10

 * postgresql://ugevpyllcosyvj:***@ec2-44-198-100-81.compute-1.amazonaws.com:5432/db4vmf7pt6qmqj
77 rows affected.


[('HL Mountain Tire', 1, 39),
 ('Patch Kit/8 Patches', 1, 34),
 ('Mountain Tire Tube', 1, 28),
 ('Road-150 Red, 62', 1, 26),
 ('Road-150 Red, 48', 1, 21),
 ('Road-150 Red, 56', 1, 19),
 ('Road-150 Red, 44', 1, 17),
 ('Fender Set - Mountain', 1, 16),
 ('Road-150 Red, 52', 1, 16),
 ('Water Bottle - 30 oz.', 1, 13),
 ('HL Mountain Tire', 4, 48),
 ('Patch Kit/8 Patches', 4, 42),
 ('Fender Set - Mountain', 4, 37),
 ('Road-150 Red, 48', 4, 37),
 ('Mountain Tire Tube', 4, 36),
 ('Road-150 Red, 56', 4, 31),
 ('Road-150 Red, 62', 4, 30),
 ('Road-150 Red, 52', 4, 29),
 ('Road-150 Red, 44', 4, 23),
 ('Road Tire Tube', 4, 23),
 ('Patch Kit/8 Patches', 5, 1),
 ('Mountain Bottle Cage', 5, 1),
 ('Classic Vest, L', 5, 1),
 ('Water Bottle - 30 oz.', 5, 1),
 ('Mountain-200 Silver, 38', 5, 1),
 ('Road Tire Tube', 6, 83),
 ('Patch Kit/8 Patches', 6, 79),
 ('Mountain Tire Tube', 6, 54),
 ('Touring Tire Tube', 6, 49),
 ('Sport-100 Helmet, Black', 6, 34),
 ('ML Road Tire', 6, 34),
 ('Water Bottle - 30 oz.', 6, 32),
 ('HL Road Tire', 6, 32),
 ('Sport-100 Helmet, Blue', 6, 32),
 ('LL Mountain Tire', 6, 30),
 ('Water Bottle - 30 oz.', 7, 58),
 ('Road Bottle Cage', 7, 39),
 ('Sport-100 Helmet, Blue', 7, 36),
 ('Mountain Bottle Cage', 7, 32),
 ('Mountain-200 Black, 42', 7, 29),
 ('Sport-100 Helmet, Red', 7, 28),
 ('Mountain-200 Black, 46', 7, 27),
 ('Mountain-200 Silver, 46', 7, 24),
 ('Patch Kit/8 Patches', 7, 22),
 ('Mountain-200 Silver, 38', 7, 21),
 ('Water Bottle - 30 oz.', 8, 52),
 ('Mountain Bottle Cage', 8, 40),
 ('AWC Logo Cap', 8, 34),
 ('Mountain-200 Silver, 46', 8, 27),
 ('Sport-100 Helmet, Black', 8, 26),
 ('Sport-100 Helmet, Red', 8, 25),
 ('Road Bottle Cage', 8, 24),
 ('Sport-100 Helmet, Blue', 8, 24),
 ('Fender Set - Mountain', 8, 18),
 ('Mountain-200 Silver, 38', 8, 17),
 ('Mountain-200 Black, 46', 8, 17),
 ('Water Bottle - 30 oz.', 9, 124),
 ('Mountain Bottle Cage', 9, 76),
 ('Mountain-200 Black, 42', 9, 71),
 ('Mountain-200 Silver, 38', 9, 69),
 ('Sport-100 Helmet, Red', 9, 68),
 ('Fender Set - Mountain', 9, 66),
 ('Road Bottle Cage', 9, 63),
 ('Sport-100 Helmet, Blue', 9, 63),
 ('Mountain-200 Silver, 42', 9, 58),
 ('Patch Kit/8 Patches', 9, 57),
 ('Water Bottle - 30 oz.', 10, 64),
 ('Sport-100 Helmet, Blue', 10, 42),
 ('Sport-100 Helmet, Red', 10, 41),
 ('AWC Logo Cap', 10, 41),
 ('Mountain Bottle Cage', 10, 38),
 ('Road Bottle Cage', 10, 36),
 ('Sport-100 Helmet, Black', 10, 23),
 ('Mountain-200 Black, 42', 10, 22),
 ('Mountain-200 Silver, 42', 10, 21),
 ('Touring Tire', 10, 20),
 ('Mountain-200 Black, 46', 10, 20)]

Tabel di atas menunjukkan 10 produk paling terlaris di setiap cabang BeeCycle. <br>
HL Mountain Tire adalah produk paling laris di cabang 1 dan 4 terjual sebanyak 39 dan 48 produk secara berurutan. <br>
Road Tire Tube adalah produk paling laris di cabang 6 terjual sebanyak 83 produk. <br>
Dan terakhir Water Bottle - 30 oz, adalah produk paling laris di banyak cabang seperti 7,8,9 dan 10 terjual sebanyak 58,52,124 dan 64 produk secara berurutan.


### Question 4

Objecive Analysis : <br>
Campaign selanjutnya BeeCycle ingin meningkatkan pendapatan perusahaan, dengan cara memperbanyak produksi dan penjualan jenis sepeda yang menghasilkan profit paling besar, terutama di wilayah United Kingdom. Agar tercapai target pasar yang sesuai, Perusahaan ingin me highlight berapa rentang umur pembeli dan jenis sepeda yang paling profitable

Business Question:
Tipe sepeda apa yang paling banyak menghasilkan revenue dalam 3 tahun terakhir di wilayah United Kingdom jika berdasarkan rentang umur pembeli??

Query Building For The Question:

In [7]:
%%sql
with CTE3 as
(select extract('year' from order_date) years,
sub_category,
case when (date_part('year', current_date) - date_part('year', birthdate)) <= 20 then 'Group <=20' 
when (date_part('year', current_date) - date_part('year', birthdate)) between 21 and 40 then 'Group 21 - 40'
when (date_part('year', current_date) - date_part('year', birthdate)) between 41 and 60 then 'Group 41 - 60'
else 'Group > 60' end Age_category,
region,
sum(totalprice_rupiah) revenue,
rank() over (partition by extract('year' from order_date) order by sum(totalprice_rupiah) desc) sort
from fact_sales
left join dim_customer using (customer_id)
left join dim_product using (product_id)
left join dim_territory using(territory_id)
where region in ('United Kingdom')
and category in ('Bikes')
group by 1,2,3,4
order by years,revenue desc)

select years,sub_category,age_category,region,revenue
from CTE3 
where sort <= 3
and years in (2019,2020,2021)

 * postgresql://ugevpyllcosyvj:***@ec2-44-198-100-81.compute-1.amazonaws.com:5432/db4vmf7pt6qmqj
9 rows affected.


[(2019.0, 'Mountain Bikes', 'Group 21 - 40', 'United Kingdom', Decimal('1007072049')),
 (2019.0, 'Mountain Bikes', 'Group 41 - 60', 'United Kingdom', Decimal('811532248')),
 (2019.0, 'Touring Bikes', 'Group 21 - 40', 'United Kingdom', Decimal('340404960')),
 (2020.0, 'Touring Bikes', 'Group 21 - 40', 'United Kingdom', Decimal('573123180')),
 (2020.0, 'Mountain Bikes', 'Group 21 - 40', 'United Kingdom', Decimal('355717180')),
 (2020.0, 'Touring Bikes', 'Group 41 - 60', 'United Kingdom', Decimal('327813780')),
 (2021.0, 'Touring Bikes', 'Group 41 - 60', 'United Kingdom', Decimal('198063600')),
 (2021.0, 'Mountain Bikes', 'Group 21 - 40', 'United Kingdom', Decimal('129219440')),
 (2021.0, 'Touring Bikes', 'Group 21 - 40', 'United Kingdom', Decimal('120916740'))]

Terlihat di 3 tahun terakhir Mountain dan Touring bike menjadi tipe sepeda yang paling profitable di United Kingdom, juga kebanyakan pembeli terbanyak adalah orang dewasa umur 21 sampai 60 tahun. 

Actionable Knowledge: <br>
Strategi BeeCycle untuk campaign selanjutnya di wilayah United Kingdom, salah satunya bisa meningkatkan produksi dan penjualan sepeda tipe Mountain dan Touring dengan standard ukuran yang cocok untuk orang dewasa. BeeCycle juga bisa memperbanyak dan menjual produk produk seperti aksesoris dan komponen yang dapat menunjang pengalaman bersepeda untuk tipe tersebut. 